database/apikeys.go

```diff
@@ -0,0 +1,100 @@
+package database
+
+import (
+	"context"
+	"encoding/base64"
+	"time"
+
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/NebulousLabs/fastrand"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+/**
+API keys are authentication tokens generated by users. They do not expire, thus
+allowing users to use them for a long time and to embed them in apps and on
+machines. API keys can be revoked when they are no longer needed or if they get
+compromised. This is done by deleting them from this service.
+*/
+
+type (
+	// APIKey is a non-expiring authentication token generated on user demand.
+	APIKey struct {
+		ID     primitive.ObjectID `bson:"_id,omitempty" json:"id"`
+		UserID primitive.ObjectID `bson:"user_id" json:"-"`
+		// Key is a base64URL-encoded []byte with length PubKeySize
+		Key       string    `bson:"key" json:"key"`
+		CreatedAt time.Time `bson:"created_at" json:"createdAt"`
+	}
+)
+
+// APIKeyCreate creates a new API key.
+func (db *DB) APIKeyCreate(ctx context.Context, user User) (*APIKey, error) {
+	if user.ID.IsZero() {
+		return nil, errors.New("invalid user")
+	}
+	ak := APIKey{
+		UserID:    user.ID,
+		Key:       base64.URLEncoding.EncodeToString(fastrand.Bytes(PubKeySize)),
```

Comment:
Let's add an index that's unique over `user_id` and `key`, that makes sense I think because your deletes do `DeleteOne`. I know it'll "never" happen but one day we'll manually insert these, or migrate these, or whatever, and we'll forget about the constraint.

Reply:
I later found out you have it but on just the `key`, which is fine. Leaving the comment here in case I'm wrong and it was an oversight on your part, but I don't think that's the case, having it on `key` is the way to go.